File: Source/Plugins/Core/com.equella.serverbase/scalasrc/com/tle/core/db/migration/NewAuditLogColumn.scala
/*
 * Copyright 2017 Apereo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tle.core.db.migration

import com.tle.core.i18n.ServerStrings
import com.tle.core.migration.MigrationResult
import io.doolse.simpledba.jdbc.JDBCIO

object NewAuditLogColumn
    extends SimpleMigration("NewAuditLogColumn",
                            2018,
                            6,
                            1,
                            ServerStrings.lookup.prefix("mig.auditcol")) {
  def migration(progress: MigrationResult, schemaMigration: DBSchemaMigration): JDBCIO[Unit] = {
    schemaMigration.addColumns(schemaMigration.auditLogNewColumns, progress)
  }
}

object NewViewCountTables
    extends SimpleMigration("NewViewCountTables",
                            2018,
                            6,
                            26,
                            ServerStrings.lookup.prefix("mig.viewcount")) {
  def migration(progress: MigrationResult, schemaMigration: DBSchemaMigration): JDBCIO[Unit] = {
    schemaMigration.addTables(schemaMigration.viewCountTables, progress)
  }
}

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/institution/migration/v66/NewAuditLogColumn.java
/*
 * Licensed to The Apereo Foundation under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * The Apereo Foundation licenses this file to you under the Apache License,
 * Version 2.0, (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tle.core.institution.migration.v66;

import com.tle.core.guice.Bind;
import com.tle.core.hibernate.impl.HibernateMigrationHelper;
import com.tle.core.migration.AbstractHibernateSchemaMigration;
import com.tle.core.migration.MigrationInfo;
import com.tle.core.migration.MigrationResult;
import java.util.List;
import javax.inject.Singleton;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import org.hibernate.Session;
import org.hibernate.annotations.AttributeAccessor;
import org.hibernate.annotations.Type;

@Bind
@Singleton
public class NewAuditLogColumn extends AbstractHibernateSchemaMigration {

  @Override
  protected void executeDataMigration(
      HibernateMigrationHelper helper, MigrationResult result, Session session) throws Exception {}

  @Override
  protected int countDataMigrations(HibernateMigrationHelper helper, Session session) {
    return 1;
  }

  @Override
  public boolean isBackwardsCompatible() {
    return true;
  }

  @Override
  protected List<String> getDropModifySql(HibernateMigrationHelper helper) {
    return null;
  }

  @Override
  protected List<String> getAddSql(HibernateMigrationHelper helper) {
    return helper.getAddColumnsSQL("audit_log_entry", "meta");
  }

  @Override
  protected Class<?>[] getDomainClasses() {
    return new Class<?>[] {FakeAuditLogEntry.class};
  }

  @Override
  public MigrationInfo createMigrationInfo() {
    return new MigrationInfo("com.tle.core.entity.services.migration.v66.auditlogentry.meta");
  }

  @Entity(name = "AuditLogEntry")
  @AttributeAccessor("field")
  public static class FakeAuditLogEntry {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    long id;

    @Type(type = "json")
    private String meta;
  }
}
